### PR TITLE
Fix `no-invalid-position-declaration` false positives for embedded styles

### LINKED ISSUE #8695: Fix `no-invalid-position-declaration` false positives for embedded styles
### What minimal example or steps are needed to reproduce the bug?

```html
<div style="color: red"></div>
```

### What minimal configuration is needed to reproduce the bug?

```json
{
  "overrides": [
    { "files": ["**/*.html"], "customSyntax": "postcss-html" }
  ],
  "rules": {
    "no-invalid-position-declaration": true
  }
}
```

### How did you run Stylelint?

[Demo](https://stylelint.io/demo/#N4Igxg9gJgpiBcIA8UCWA3ABAZwC4E8AbGAXgB1wJCIAneTGmKCgPiQHo10WyA7EADQgAZqmIA5AIYBbOIhgAPGQAdiAOgAWuaYUGVeogOYIQwPpkwUI6GDRqpY2CvQDa5i5mCWRYmE4SYbiAAVMHswZrahBQAugLeYACueBDSAMr4vLiSCs7eyhB4YNjYALRaOhSYAL7uce4UNInE-vRmvB7evBClqLzokoQOpQXYqLioELylsGCEkjSSE1N5uE0w7rW81XqQBqiGAGK00ksmAFbYU3qwytgm7RYUeEQwQ1l5FACMAGxqAEwAZjUAAYKAIGiAXsR3rhSnsjKU8JJeFAFswAhR5rg-LhwZDRrhimUKtFMSBsbiKHwdtUgA)

### Which Stylelint-related dependencies are you using?

```json
{
  "stylelint": "16.23.0",
  "postcss-html": "latest"
}
```

### What did you expect to happen?

No problem.

### What actually happened?

`color: red` is reported as a problem.

### Do you have a proposal to fix the bug?

_No response_

### PR DESCRIPTION
**Pull request by sw1tch3roo**
> Which issue, if any, is this issue related to?

Closes #8695

> Is there anything in the PR that needs further explanation?

No, it's self-explanatory.

### CODE REVIEW

**Comment by sw1tch3roo** [comment]
inspired by https://github.com/stylelint/stylelint/pull/8697. 
@reduckted @jeddy3 guys, what do you think about this?

**Comment by reduckted** [comment]
Is it worth adding an `accept` test case for `<style>--custom-property: blue;</style>` to show that the rule will produce a false negative for check `style` elements?

**Comment by sw1tch3roo** [comment]
@reduckted Thank you for your suggestion, IMO it **would be worth**.
```javascript
<style>--custom-property: blue;</style>
```

this way we ensure that the fix won't be repeated in future changes.

File: .changeset/yummy-streets-cheat.md [hunk 1]
```diff
@@ -0,0 +1,5 @@
+---
+"stylelint": patch
+---
+
+Fixed: `no-invalid-position-declaration` skipped in HTML style attributes
```
Comment (jeddy3) [inline .changeset/yummy-streets-cheat.md]:
```suggestion
Fixed: `no-invalid-position-declaration` false positives for embedded styles
```

For [our](https://stylelint.io/maintainer-guide/pull-requests#titles) [conventions](https://stylelint.io/maintainer-guide/issues#titles)

File: lib/rules/no-invalid-position-declaration/__tests__/index.mjs [hunk 2]
```diff
@@ -316,3 +316,45 @@ testRule({
...
+		{
+			code: '<span style="font-size: 14px; margin: 10px;"></span>',
+			description: 'HTML style attribute with multiple properties',
+		},
+		{
+			code: '<style>--custom-property: blue;</style>',
+			description: 'HTML style element with custom property',
+		},
+	],
+});
```
Comment (jeddy3) [inline lib/rules/no-invalid-position-declaration/__tests__/index.mjs]:
```suggestion
testRule({
	ruleName,
	config: true,
	customSyntax: 'postcss-html',

	accept: [
		{
			code: '<a style="color: red"></a>',
			description: 'HTML style attribute',
		},
	],

	reject: [
		{
			skip: true,
			code: '<style>color: red</style>',
			message: messages.rejected,
			line: 1,
			column: 6,
			endLine: 1,
			endColumn: 10,
		},
	],
});
```

Let's

- keep the test lean for non-standard syntax, as, ideally, we wouldn't have them
- add and skip a test for the false negatives

File: lib/rules/no-invalid-position-declaration/__tests__/index.mjs [hunk 3]
```diff
@@ -316,3 +316,45 @@ testRule({
...
+		{
+			code: 'const foo = styled.a`color: red;`',
+			description: 'CSS-in-JS template literal',
+		},
+		{
+			code: 'const bar = styled.div`--custom: blue; font-size: 16px;`',
+			description: 'CSS-in-JS with custom property',
+		},
+	],
+});
```
Comment (jeddy3) [inline lib/rules/no-invalid-position-declaration/__tests__/index.mjs]:
```suggestion
testRule({
	ruleName,
	config: true,
	customSyntax: naiveCssInJs,

	accept: [
		{
			code: 'const foo = styled.a`color: red;`',
			description: 'CSS-in-JS template literal',
		},
	],
});
```

Let's use the naive syntax rather than add a new dev dependency.

```js
import { stripIndent } from 'common-tags';

import naiveCssInJs from '../../../__tests__/fixtures/postcss-naive-css-in-js.cjs';
import rule from '../index.mjs';
```

File: lib/utils/typeGuards.mjs [hunk 4]
```diff
@@ -49,6 +49,33 @@ export function isDocument(node) {
...
+		) {
+			return true;
+		}
+
+		current = current.parent;
+	}
+
+	return false;
+}
+
```
Comment (jeddy3) [inline lib/utils/typeGuards.mjs]:
```suggestion
/**
 * @param {import('postcss').Node} node
 * @returns {boolean}
 */
export function isInDocument({ parent }) {
	return Boolean(parent && (isDocument(parent) || 'document' in parent));
}

```

Will this simpler version, which doesn't traverse up the tree, suffice since we're only checking declarations in the root?

File: package.json [hunk 5]
```diff
@@ -211,6 +211,7 @@
     "postcss-less": "^6.0.0",
     "postcss-sass": "^0.5.0",
     "postcss-scss": "^4.0.9",
+    "postcss-styled-syntax": "^0.7.1",
```
Comment (jeddy3) [inline package.json]:
```suggestion
```

**Review by jeddy3 — CHANGES_REQUESTED** [review]
@sw1tch3roo Thank you for continuing to work on a fix.

I've requested some changes and asked a question.

File: lib/utils/typeGuards.mjs (hunk 4 above)
Comment (sw1tch3roo) [reply lib/utils/typeGuards.mjs]: Yep, I think the simpler version does the job just fine — especially for this specific use case.

it turns out PostCSS builds the AST that way by default, so there's no need to walk up the tree — the info we need is already in the immediate parent (e.g. `decl.parent.document` is already there if it's from an inline style).

So yeah, the simplified version with just `parent && (isDocument(parent) || 'document' in parent)` should be totally fine.

Happy to revert to the full traversal if we ever hit edge cases, but for now this should be solid.

**Review by jeddy3 — APPROVED** [review]
Thank you for making the changes. LGTM.

File: lib/rules/no-invalid-position-declaration/index.mjs [hunk 6]
```diff
@@ -41,9 +44,7 @@ const rule = (primary) => {
 					const parentRule = findNodeUpToRoot(decl, ({ type }) => type === 'rule');
 
 					if (parentRule) return;
-				} else {
-					return;
-				}
+				} else return;
```
Comment (romainmenke) [inline lib/rules/no-invalid-position-declaration/index.mjs]: This seems unrelated to the issue.
Maybe better to keep it as it was so that the git history is more meaningful?

**Review by romainmenke — COMMENTED** [review]
Overal change looks good! Thank you
I've left a few suggestions, please take a look when you have time 🙇

File: lib/utils/typeGuards.mjs [hunk 7]
```diff
@@ -49,6 +49,14 @@ export function isDocument(node) {
 	return node.type === 'document';
 }
 
+/**
+ * @param {import('postcss').Node} node
+ * @returns {boolean}
+ */
+export function isInDocument({ parent }) {
+	return Boolean(parent && (isDocument(parent) || 'document' in parent));
+}
```
Comment (romainmenke) [inline lib/utils/typeGuards.mjs]: This is not a type guard.

Maybe better to keep this function local to the rule?

**Review by romainmenke — COMMENTED** [review]
(this comment went missing in the first review)

File: lib/rules/no-invalid-position-declaration/__tests__/index.mjs [hunk 8]
```diff
@@ -316,3 +317,38 @@ testRule({
...
+
+testRule({
+	ruleName,
+	config: true,
+	customSyntax: naiveCssInJs,
+	accept: [
+		{
+			code: 'const foo = styled.a`color: red;`',
```
Comment (ybiquitous) [inline lib/rules/no-invalid-position-declaration/__tests__/index.mjs]: [suggestion] This test code seems to have no sense for `naiveCssInJs`. How about fixing it like so?

https://github.com/stylelint/stylelint/blob/b17f5a421fbba4867db9667bf46934450962f639/lib/__tests__/fixtures/postcss-naive-css-in-js.cjs#L16-L17

```suggestion
			code: 'css` color: red; `;',
```

File: lib/utils/typeGuards.mjs [hunk 9]
```diff
@@ -49,6 +49,14 @@ export function isDocument(node) {
 	return node.type === 'document';
 }
 
+/**
+ * @param {import('postcss').Node} node
+ * @returns {boolean}
+ */
+export function isInDocument({ parent }) {
+	return Boolean(parent && (isDocument(parent) || 'document' in parent));
```
Comment (ybiquitous) [inline lib/utils/typeGuards.mjs]: [suggestion] Since this `isInDocument()` is not a *type guard*, I'd like to extract it as a new file, such as `lib/utils/isisInDocument.mjs`.

---

[question] By the way, may a `Node` have a `document` property? 🤔
I cannot find the property in the [PostCSS API](https://postcss.org/api/).

File: lib/rules/no-invalid-position-declaration/__tests__/index.mjs (hunk 8 above)
Comment (ybiquitous) [reply lib/rules/no-invalid-position-declaration/__tests__/index.mjs]: Just opened #8715 to prevent such a case.

File: lib/utils/typeGuards.mjs (hunk 7 above)
Comment (sw1tch3roo) [reply lib/utils/typeGuards.mjs]: Can we move it to `lib/utils`?

Comment (romainmenke) [reply lib/utils/typeGuards.mjs]: Sure! :)

File: lib/rules/no-invalid-position-declaration/index.mjs (hunk 6 above)
Comment (sw1tch3roo) [reply lib/rules/no-invalid-position-declaration/index.mjs]: I agree. then I suggest doing autofix, the [issue](https://github.com/stylelint/eslint-config-stylelint/issues/346) is open

File: lib/utils/typeGuards.mjs (hunk 4 above)
Comment (sw1tch3roo) [reply lib/utils/typeGuards.mjs]: @jeddy3 @ybiquitous we initially thought the simplified version would be enough (i.e. just checking the immediate parent), but after adding the css-in-js test with `naiveCssInJs`, it started failing — turns out the Document node is higher up in the AST (e.g. `Document` → `Root` → `Decl`), so we do need to walk up the tree after all...

File: lib/utils/typeGuards.mjs (hunk 9 above)
Comment (sw1tch3roo) [reply lib/utils/typeGuards.mjs]: Hey @ybiquitous 

as far as I've googled, **yes – but only in specific cases.**

PostCSS added an optional `document` property on **Input nodes** (used for HTML parsing or CSS‑in‑JS), not on the general `Node` interface. For example:

```js
if ('document' in node && node.document) {
  // means this decl was parsed from an HTML <style> or JS template
}
```

But for most CSS nodes (e.g. `Rule`, `AtRule`, `Decl`), `document` is **undefined** – those types don’t include that property.

So i.e., *only nodes coming from a `postcss-html` or CSS‑in‑JS context keep a `.document` backlink* – whereas a plain AST node from a CSS file doesn’t get one.

Comment (ybiquitous) [reply lib/utils/typeGuards.mjs]: Honestly, I don't think we should rely on the `document` property that is unofficial. Instead, I guess a root node parsed by `postcss-html` should have a `parent` property as a `Document` node because the `Root#document` property is defined below:

```ts
parent: Document | undefined
```

Ref https://github.com/postcss/postcss/blob/3f51dafbbf0e00655f9de13debb77dbdae872bf6/lib/root.d.ts#L58

```js
const postcssHtml = require('postcss-html');
const code = '<a style="color: red"></a>';

postcssHtml.parse(code).type
//=> 'document'
postcssHtml.parse(code).first.type
//=> 'root'
postcssHtml.parse(code).first.parent
//=> undefined (but it should be a document)
```

cc @ota-meshi

Comment (ybiquitous) [reply lib/utils/typeGuards.mjs]: For now, I'm okay that `isInDocument()` relies on the unofficial `document` property, but I'd suggest:

- Leave a comment that describes the code relies on the unofficial `document` property, e.g., from `postcss-html` etc.
- Add a few test cases using `postcss-html` for `isInDocument()`

---

In general, we shouldn't depend on the unofficial behavior of 3rd-party PostCSS parsers, as the current Stylelint prefers only the CSS standard syntax.

File: lib/utils/typeGuards.mjs (hunk 4 above)
Comment (ybiquitous) [reply lib/utils/typeGuards.mjs]: Understood.

FYI, refer to https://github.com/stylelint/stylelint/pull/8701#discussion_r2252856667

If `root.parent === document` was always satisfied, `isInDocument()` would be much simpler.

File: lib/utils/typeGuards.mjs (hunk 9 above)
Comment (sw1tch3roo) [reply lib/utils/typeGuards.mjs]: I've added some tests and left a comment about `document` property in `isInDocument` util, can you check it?

File: lib/utils/isInDocument.mjs [hunk 10]
```diff
@@ -0,0 +1,27 @@
+import { isDocument } from './typeGuards.mjs';
```
Comment (ybiquitous) [inline lib/utils/isInDocument.mjs:3]:
```suggestion
import { Node } from 'postcss';

import { isDocument } from './typeGuards.mjs';
```

File: lib/utils/isInDocument.mjs [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+
+/**
+ * @param {import('postcss').Node} node
```
Comment (ybiquitous) [inline lib/utils/isInDocument.mjs]:
```suggestion
 * @param {Node} node
```

File: lib/utils/isInDocument.mjs [hunk 12]
```diff
@@ -0,0 +1,27 @@
...
+	let current = node;
+
+	while (current) {
+		if (isDocument(current)) return true;
+
+		// Check for unofficial 'document' property from parsers like postcss-html
+		if (
+			'document' in current &&
+			current.document &&
+			isDocument(/** @type {import('postcss').Document} */ (current.document))
```
Comment (ybiquitous) [inline lib/utils/isInDocument.mjs]:
```suggestion
			current.document instanceof Node &&
			isDocument(current.document)
```

**Comment by sw1tch3roo** [comment]
> @sw1tch3roo Thanks for addressing my reviews.
> 
> Overall LGTM, but let me suggest the last refactoring in `isInDocument()`. I believe using `instanceof Node` is a bit safer and straightforward than type casting with `@type`.

it's a good catch, thanks

**Comment by ybiquitous** [comment]
@jeddy3 may want one more review since commits were added after his approval.

**Review by jeddy3 — APPROVED** [review]
LGTM, thank you!

(And thanks for catching all the things I missed in my first review 😅)
